cilium/cmd/bpf_ct_list_test.go

```diff
@@ -0,0 +1,194 @@
+	ctKey4   = ctmap.CtKey4{
+		TupleKey4: tuple.TupleKey4{
+			DestAddr:   types.IPv4{10, 10, 10, 1},
+			SourceAddr: srcAddr4,
+			DestPort:   byteorder.HostToNetwork(uint16(80)).(uint16),
+			SourcePort: byteorder.HostToNetwork(uint16(13579)).(uint16),
+			NextHeader: 6,
+			Flags:      123,
+		},
+	}
+	srcAddr6 = types.IPv6{1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 121, 98, 219, 61}
+	ctKey6   = ctmap.CtKey6{
+		TupleKey6: tuple.TupleKey6{
+			DestAddr:   types.IPv6{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+			SourceAddr: srcAddr6,
+			DestPort:   byteorder.HostToNetwork(uint16(443)).(uint16),
+			SourcePort: byteorder.HostToNetwork(uint16(7878)).(uint16),
+			NextHeader: 17,
+			Flags:      31,
+		},
+	}
+	ctValue = ctmap.CtEntry{
+		RxPackets:        1,
+		RxBytes:          512,
+		TxPackets:        4,
+		TxBytes:          2048,
+		Lifetime:         12345,
+		Flags:            3,
+		RevNAT:           byteorder.HostToNetwork(uint16(27)).(uint16),
+		TxFlagsSeen:      88,
+		RxFlagsSeen:      99,
+		SourceSecurityID: 6789,
+		LastTxReport:     0,
+		LastRxReport:     7777,
+	}
+)
+
+type ctRecord4 struct {
+	MapKey   tuple.TupleKey4
+	MapValue ctmap.CtEntry
+}
+
+type ctRecord6 struct {
+	MapKey   tuple.TupleKey6
+	MapValue ctmap.CtEntry
+}
+
+func dumpAndRead(maps []ctmap.CtMap, c *C) string {
+	// dumpCt() prints to standard output. Let's redirect it to a pipe, and
+	// read the dump from there.
+	stdout := os.Stdout
+	readEnd, writeEnd, err := os.Pipe()
+	c.Assert(err, IsNil, Commentf("failed to create pipe: '%s'", err))
+	os.Stdout = writeEnd
+	defer func() { os.Stdout = stdout }()
+
+	command.ForceJSON()
+	dumpCt("", maps)
+
+	channel := make(chan string)
+	go func() {
+		var buf bytes.Buffer
+		_, err = io.Copy(&buf, readEnd)
+		channel <- buf.String()
+	}()
+
+	writeEnd.Close()
+	// Even though we have a defer, restore os.Stdout already if we can
+	// (for the assert)
+	os.Stdout = stdout
+	rawDump := <-channel
+	c.Assert(err, IsNil, Commentf("failed to read data: '%s'", err))
+
+	return rawDump
+}
+
+func (s *BPFCtListSuite) TestDumpCt4(c *C) {
+
+	ctMaps := [2]ctmap.CtMap{
+		ctmap.NewCtMockMap(
+			[]ctmap.CtMockMapRecord{
+				{
+					Key:   &ctKey4,
+					Value: &ctValue,
+				},
+				{
+					Key:   &ctKey4,
+					Value: &ctValue,
+				},
+			},
+		),
+		ctmap.NewCtMockMap(
+			[]ctmap.CtMockMapRecord{
+				{
+					Key:   &ctKey4,
+					Value: &ctValue,
+				},
+			},
+		),
+	}
+
+	rawDump := dumpAndRead(ctMaps[:], c)
+
+	var ctDump []ctRecord4
+	err := json.Unmarshal([]byte(rawDump), &ctDump)
+	c.Assert(err, IsNil, Commentf("invalid JSON output: '%s', '%s'", err, rawDump))
+
+	c.Assert(ctDump[0].MapKey.SourceAddr, Equals, srcAddr4)
+	c.Assert(ctDump[0].MapKey.DestPort, Equals, uint16(80))
+	c.Assert(ctDump[0].MapKey.NextHeader, Equals, u8proto.U8proto(6))
+	c.Assert(ctDump[0].MapValue.RevNAT, Equals, uint16(27))
+	c.Assert(ctDump[0].MapValue.LastRxReport, Equals, uint32(7777))
```

Comment:
Any chance you could turn these lines into a `checker.DeepEquals` `Check()`?

Reply:
Probably, what's the difference exactly?

Reply:
`checker.DeepEquals` is a pretty-printer that provides diff-style comparison of observed vs. expected output. It makes interpreting the tests much easier by printing the failure inline, rather than just getting something like "Assert failed on line 42".

Reply:
Plus, compared to simple `Equals`, you would only need a single line to check the whole object.

Reply:
Ah, so I can compare directly the whole entry? Sounds good, thanks! Will do. And thanks for the explanations.

Reply:
So that would be:
```
    c.Assert(ctDump[0].MapKey, checker.DeepEquals,
        ctMaps[0].(*ctmap.CtMockMap).Entries[0].Key.(*ctmap.CtKey4).TupleKey4)
```

But the values aren't actually equal:
```
 tuple.TupleKey4{
     DestAddr:   {0xa, 0xa, 0xa, 0x1},
     SourceAddr: {0xa, 0xa, 0xa, 0x2},
-    DestPort:   0x50,
-    SourcePort: 0x350b,
+    DestPort:   0x5000,
+    SourcePort: 0xb35,
     NextHeader: 0x6,
     Flags:      0x7b,
 }
```

Because I chose to `ntohs()` the fields that were in network byte order for readability. Which makes me wonder if this is a good choice. Do we want them readable in the JSON, at the expense of JSON entry not being a 100% exact mirror of the internal struct, or do we want parity? (If the JSON is processed at some point and we try to map it to internal structs, parity is probably better? And surely for readability there is plain output anyway?)

Reply:
I'd go for 100% exact mirror. We could also override the `Unmarshal()` function, but it's really not worth it here...